### ISSUE TITLE
[css-masking-1] Let SVGClipPathElement inherit from SVGGraphicsElement

### ISSUE THREAD

**Issue by SebastianZ** [w3c/fxtf-drafts#28]
As `<clipPath>` is grouping graphics elements similarily to `<g>`, its `SVGClipPathElement` interface should also inherit from `SVGGraphicsElement` like `SVGGElement` does.

Sebastian

**Comment by AmeliaBR** [comment]
SVGGraphicsElement is used for elements that are directly part of [the rendering tree](https://svgwg.org/svg2-draft/render.html#RenderingTree). None of the other SVG graphics effects elements (filters, patterns, gradients) inherit from SVGGraphicsElement.

Was there a particular method defined for SVGGraphicsElement that you wanted to have available on `clipPath` or `mask`?

**Comment by SebastianZ** [comment]
Yes, I thought `getBBox()` would be useful to have available on them. One use case for it would be to have the highlighter of browser developer tools available `<clipPath>` and ´<mask>` elements.

Sebastian

**Comment by AmeliaBR** [comment]
I'm not sure how the highlighter would work, since you never know what coordinate system the elements are in; it depends on the element using the effect.

`getBBox()` could work because it doesn't actually depend on a parent coordinate system, only on the child content.  But the other methods currently defined for graphics elements (cumulative transform matrices) don't make sense.

**Comment by SebastianZ** [comment]
> I'm not sure how the highlighter would work, since you never know what coordinate system the 
> elements are in; it depends on the element using the effect.

There could actually be two highlighters, one for the mask or clip path itself without being applied and one for when they are used (which may then highlight all usages).

> `getBBox()` could work because it doesn't actually depend on a parent coordinate system, only on the child content. But the other methods currently defined for graphics elements (cumulative transform matrices) don't make sense.

Agree. So I see three solutions:

1. Add `getBBox()` to `SVGClipPathElement` and `SVGMaskElement` rather than letting them inherit from `SVGGraphicsElement`
2. Let `SVGClipPathElement` and `SVGMaskElement` inherit from `SVGGraphicsElement` and define that the cumulative transform matrice methods return an invalid value for them.
3. Move `getBBox()` to a new interface and let `SVGGraphicsElement`, `SVGClipPathElement` and `SVGMaskElement` inherit from it.

I'd be fine with all three of them, but I think number 3 would be the cleanest.

Sebastian

**Comment by dirkschulze** [comment]
@SebastianZ This would not fit into Firefox's internal rendering model. I would suggest to just create a union of the content elements of a clipping path but realised that this might not work either in Firefox.

Point 2 would most definitely not work as @AmeliaBR said.

**Comment by dirkschulze** [comment]
@bzbarsky Do you have some input from Geckos point of view? I know that in the past it was not possible for Firefox to get bounds of content that does not get directly rendered. Like content from `<mask>` or `<clipPath>`.

@SebastianZ At this point I am hesitating to add this to CSS Masking 1 and would rather defer the request to CSS Masking 2. I'd like CSS Masking 1 to consolidate with current implementations.

**Comment by SebastianZ** [comment]
> @SebastianZ At this point I am hesitating to add this to CSS Masking 1 and would rather defer the request to CSS Masking 2. I'd like CSS Masking 1 to consolidate with current implementations.

Fair enough.

Sebastian

**Comment by bzbarsky** [comment]
I believe the issue is that getBBox in Gecko uses the rendering tree.  That's partly because getBBox relies on styles and until recently those were very much tied to the rendering tree in Gecko.  And even in the new style system, we optimize out style computation in various cases (like display:none)...

All of which is to say that I have no idea when Gecko will have a getBBox that does not rely on the rendering tree.
